follower/execution/follower.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package execution
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+const cacheSize = 1024
+
+type blockReader interface {
+	Notify() <-chan string
+	Read(filename string) ([]byte, error)
+}
+
+type Follower struct {
+	log zerolog.Logger
+
+	blocks blockReader
+	codec  dps.Codec
+	db     *badger.DB
+
+	data    map[uint64]*BlockData
+	current *BlockData
+	height  uint64
+	index   int
+
+	stop chan struct{}
+}
+
+func New(log zerolog.Logger, blocks blockReader, codec dps.Codec, db *badger.DB) *Follower {
+	f := Follower{
+		log: log,
+
+		blocks: blocks,
+		codec:  codec,
+		db:     db,
+
+		height: math.MaxUint64,
+		data:   make(map[uint64]*BlockData, cacheSize),
+
+		stop: make(chan struct{}),
+	}
+
+	return &f
+}
+
+// Run launches the execution follower.
+func (f *Follower) Run() error {
+	tick := time.NewTicker(1 * time.Second)
+	defer tick.Stop()
+
+	for range tick.C {
```

Comment:
I just find this inelegant. There is no reason to check every second (or any other set amount of time). Network I/O could easily be higher than one second at any one point. With this approach we will 1) never get _optimal_ latency (within the constraints of the file-based approach) 2) never make _optimal_ amount of requests.